cpp/src/interop/to_arrow_device.cu

```diff
@@ -586,6 +378,207 @@ int dispatch_to_arrow_device::operator()<cudf::dictionary32>(cudf::column&& colu
   return NANOARROW_OK;
 }
 
+struct dispatch_to_arrow_device_view {
+  cudf::column_view column;
+  rmm::cuda_stream_view stream;
+  rmm::mr::device_memory_resource* mr;
+
+  template <typename T, CUDF_ENABLE_IF(not is_rep_layout_compatible<T>())>
+  int operator()(ArrowArray*) const
+  {
+    CUDF_FAIL("Unsupported type for to_arrow_device", cudf::data_type_error);
+  }
+
+  template <typename T, CUDF_ENABLE_IF(is_rep_layout_compatible<T>())>
+  int operator()(ArrowArray* out) const
+  {
+    nanoarrow::UniqueArray tmp;
+
+    auto const storage_type = id_to_arrow_storage_type(column.type().id());
+    NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), storage_type, column));
+    NANOARROW_RETURN_NOT_OK(set_null_mask(column, tmp.get()));
+    NANOARROW_RETURN_NOT_OK(set_view_to_buffer(column, tmp.get()));
+
+    ArrowArrayMove(tmp.get(), out);
+    return NANOARROW_OK;
+  }
+
+  int set_buffer_view(void const* in_ptr, size_t size, int64_t i, ArrowArray* out) const
+  {
+    ArrowBuffer* buf = ArrowArrayBuffer(out, i);
+    buf->size_bytes  = size;
+    auto const ptr   = reinterpret_cast<uint8_t const*>(in_ptr);
+
+    // reset the deallocator to do nothing since this is a non-owning view
+    auto no_deallocate = [](ArrowBufferAllocator*, uint8_t*, int64_t) {};
```

Comment:
Should we make this static or otherwise avoid creating a new empty lambda for every invocation (such as shifting it outside of the function definition)?

Reply:
I believe the deallocator makes a copy of it anyway so a static here would not save anything.

Reply:
Should we inline the declaration in the ArrowBufferDeallocator call, then? I don't know off the top of my head if this matters for lambdas or functors.

Reply:
I prefer how this looks myself since clang-format already breaks up the single line of code into 2 lines.
I don't think it matters to the compiler so I'd prefer readability.

Reply:
It looks like I was slow and you went and ahead and inlined as I requested. I'm fine either way; I'm assuming that you are too since you made the change, but feel free to revert if you want.

Reply:
Thanks. I liked that it is now 2 lines instead of 3.  :smirk: